[Doc Repo] Enhancements to readability and logging for edit/upload ajax file

johnsaigle opened the pull request:
This pull request touches up the doc repo to enhance its readability, error reporting, and performance.

I have self-reviewed to provide feedback on specific points. 

### To Test

* Everything should work the same as before. Try basic doc repo functions and make sure nothing is broken.

johnsaigle commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -2,7 +2,7 @@
   * Document_repository module
   *
-  * PHP Version 5
+  * PHP Version 7
```
Used `??` operator which requires >= 7.0

johnsaigle commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +52,35 @@
+
+        $fileSize = $_FILES['file']['size'];
+        $fileName = $_FILES['file']['name'];
+        $fileType = pathinfo($_FILES['file']['name'], PATHINFO_EXTENSION);
```
See: https://stackoverflow.com/a/4591546
See comments: https://stackoverflow.com/a/5427229

johnsaigle commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +52,35 @@
     || $userSingleton->hasPermission('document_repository_delete')
 ) {
     if ($action == 'upload') {
-        $puser      = $_POST['user'];
```
Unnecessary: this is just the username. Used config setting instead.

johnsaigle commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +52,35 @@
-        $fileBase  = $puser . "/" . $fileName;
-
-        if (!file_exists($base_path . $puser)) {
-            mkdir($base_path . $puser, 0777);
```
A world-executable directory is probably unnecessary here. Using `700` instead.

ridz1208 replied:
I think 770 is what the installer does. Maybe remove the create entirely ? Not sure i like it in the code

johnsaigle replied:
Yeah that makes sense. I think the installer didn't create the directory at some point in the past.

johnsaigle replied:
This has now been changed to `770`, The install script [creates a directory](https://github.com/aces/Loris/blob/minor/tools/install.sh#L158) owned by www-data and there's no implication another user would need access to this folder -- so this should work.

johnsaigle commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -120,7 +115,7 @@
         $comments   = $_POST['commentsEdit'];
         $version    = $_POST['versionEdit'];
 
-        if (empty($category) && $category !== '0') {
```
`empty` checks for `"0"` so this is redundant.

ridz1208 replied:
This seems to be purposefully checking for a 0 string. You are changing behavior here. 

Ie if categories = 0, ur change makes no difference. But is categories ='0', the original code will evaluate to false while urs will evaluate to true. It is not redundant in this case

johnsaigle replied:
Oh I see. I read that wrong. Good catch

johnsaigle replied:
I've changed this check to cast `$category` as an int and check whether its value is >= 0, and added a note to this effect. This makes it more clear what's happening, imo.

ZainVirani replied:
I feel like a comment could have made it more clear, personally not a fan of the int cast, but also i don't care enough to request a change 😃

johnsaigle replied:
@ZainVirani I decided to use `is_numeric` so that this would be more clear

johnsaigle replied:
Added a comment so we get the best of all possible worlds

johnsaigle commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -89,26 +90,20 @@
             $uploadNotifier->notify($msg_data);
 
-            $header = "Location:".
-                      " $baseURL/document_repository/?uploadSuccess=true";
```
This URL param isn't used anywhere.

johnsaigle commented on modules/document_repository/templates/menu_document_repository.tpl:
```diff
@@ -364,7 +364,6 @@
                                 <input type="text" size = "27" name="version" id="version" class="ui-corner-all form-fields form-control input-sm" />
                             </div>
                         </div>
-                        <input type="hidden" name = "user" id = "user" value = "{$User}">
```
There's no need to POST the username. We already have it in the back-end.

johnsaigle replied:
Also see https://github.com/aces/Loris/pull/3678

johnsaigle commented on modules/document_repository/templates/menu_document_repository.tpl:
```diff
@@ -364,9 +364,7 @@
                         </div>
-                        <input type="hidden" name = "user" id = "user" value = "{$User}">
                         <input type="hidden" name = "action" id = "action" value = "upload">
-                        <input type="hidden" id="MAX_FILE_SIZE" name="MAX_FILE_SIZE" value="100000000" />
```
A nice idea but this also isn't anywhere; even if it was this sort of validation on the front-end could easily be spoofed.

We should validate this on the back end at some point in the future.

ridz1208 replied:
Would it be a good place to add it here given this deletion ??

johnsaigle replied:
Do we still have a distinction between cleanup and feature pull requests? I agree we should have one but I wanted this PR to simply be an enhancement rather than what is effectively a new feature.

johnsaigle replied:
I'll create a Redmine ticket for it and issue a fix to the bugfix branch

johnsaigle replied:
@ridz1208 I believe that this sort of thing is actually handled by the `php.ini` file. I'm not sure if we have any explicit conventions with regard to max file upload size but I believe this is handled there rather than here.

PapillonMcGill replied:
@johnsaigle @ridz1208 the MAX_FILE_SIZE is effectively define in the php.ini. It might still be a good idea to check on the client side and report that the file is to big instead of waiting for the server to report and error AFTER downloading 2GB of data. A check could also be implemented on the server side before actually trying to do it. Faster for the user to see the error and prevent clogging the bandwidth.

johnsaigle replied:
@PapillonMcGill I agree. I'll create a ticket in Redmine.

HenriRabalais commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -17,32 +17,32 @@
-    "document_repository",
-    "upload"
+    'document_repository',
+    'upload'
```
YES. Take those unnecessary double quotes down!!

johnsaigle replied:
Definitely a small change but in some cases using single quotes vs. double quotes can [give a small boost in speed](https://stackoverflow.com/questions/3446216/what-is-the-difference-between-single-quoted-and-double-quoted-strings-in-php?answertab=votes#tab-top). It kind of depends on what the compiler does ultimately and probably doens't matter but.... figured I might as well at least keep this file consistent, one way or another.

johnsaigle commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +52,35 @@
-
-        if (move_uploaded_file($_FILES["file"]["tmp_name"], $target_path)) {
+        if (!move_uploaded_file($_FILES['file']['tmp_name'], $fullPath)) {
+            error_log('File upload failed for unknown reasons.');
```
Happy path refactoring

johnsaigle commented:
@ridz1208 please re0review when you have time

ZainVirani commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +52,35 @@
-        if (!file_exists($base_path . $puser)) {
-            mkdir($base_path . $puser, 0777);
+        $category   = $_POST['category']; # required
+        $site       = $_POST['site'] ?? ''; # required but allowed to be empty
```
1 lonely `??` not aligned with his family 😢

xlecours commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -17,32 +17,31 @@
 set_include_path(
     get_include_path().
-    ":../../project/libraries:../../php/libraries:"
+    ':../../project/libraries:../../php/libraries:'
```
AjaxHelper is taking care of this part. And of the following require_once

xlecours commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -17,32 +17,31 @@
 $factory = NDB_Factory::singleton();
 $baseURL = $factory->settings()->getBaseURL();
 
 $config = NDB_Config::singleton();
```
Could you add $factory->config() instead of using the singleton?

xlecours commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +51,37 @@
+
+        if (!is_writable($uploadPath)) {
+            if (file_exists($uploadPath)) {
+                error_log("Could not write to $uploadPath. Check permissions");
```
HTTP/1.1 403 Forbidden

xlecours commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +51,37 @@
+                error_log("Could not write to $uploadPath. Check permissions");
+                exit;
+            }
+            error_log(
```
Is this necessary?

johnsaigle replied:
I thought it might be good to log when a new folder in the file system is being created but I can remove this if you think it's too verbose.

xlecours replied:
I just don't think anyone would go into error log to check that.
But I think we should have a log directory though to put those kind of things

xlecours commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +51,37 @@
-        if (move_uploaded_file($_FILES["file"]["tmp_name"], $target_path)) {
+        if (!move_uploaded_file($_FILES['file']['tmp_name'], $fullPath)) {
+            error_log('File upload failed for unknown reasons.');
+            echo('ERROR: Could not upload file. Contact your administrator');
```
HTTP/1.1 500 Internal Server Error

xlecours commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -89,26 +91,20 @@
-
-        } else {
-            echo "There was an error uploading the file";
+            header('Location:' . $baseURL . '/document_repository/');
```
please add `HTTP/1.1 303 See Other` before
Otherwise, This  implicitly return a 302 (see php doc http://php.net/manual/en/function.header.php and https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302)

xlecours commented:
Approving. Since Travis do not complain I assume all the javascript can handle the possible errors

ZainVirani commented:
I'm getting lots of undefined index (openAccordian, filtered) and undefined variable (fileCategories) in my error log

johnsaigle commented:
@ZainVirani yeah me too. These are `Notice` level rather than warning or error, which is what this PR addresses.

ZainVirani commented:
Document upload failed when trying to upload test.txt under category Text > Test

```
[Mon May 07 12:41:31.071195 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP Warning:  PDOStatement::execute(): SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'For_site' at row 1 in /var/www/LORIS/Loris/php/libraries/Database.class.inc on line 339, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.071348 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP Stack trace:, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.071586 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP   1. {main}() /var/www/LORIS/Loris/htdocs/AjaxHelper.php:0, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.071668 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP   2. require() /var/www/LORIS/Loris/htdocs/AjaxHelper.php:104, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.071984 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP   3. Database->insert() /var/www/LORIS/Loris/modules/document_repository/ajax/documentEditUpload.php:94, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.072180 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP   4. Database->_realinsert() /var/www/LORIS/Loris/php/libraries/Database.class.inc:243, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.072363 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP   5. PDOStatement->execute() /var/www/LORIS/Loris/php/libraries/Database.class.inc:339, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.078362 2018] [:error] [pid 15077] [client 142.157.144.81:54476] Could not execute INSERT INTO document_repository SET `File_category`=:File_category,`For_site`=:For_site,`comments`=:comments,`version`=:version,`File_name`=:File_name,`File_size`=:File_size,`Data_dir`=:Data_dir,`uploaded_by`=:uploaded_by,`Instrument`=:Instrument,`PSCID`=:PSCID,`visitLabel`=:visitLabel,`File_type`=:File_type. Stack trace#0 /var/www/LORIS/Loris/php/libraries/Database.class.inc(243): Database->_realinsert('document_reposi...', Array, true)\n#1 /var/www/LORIS/Loris/modules/document_repository/ajax/documentEditUpload.php(94): Database->insert('document_reposi...', Array)\n#2 /var/www/LORIS/Loris/htdocs/AjaxHelper.php(104): require('/var/www/LORIS/...')\n#3 {main}, referer: http://132.206.37.24/document_repository/?reset=true
[Mon May 07 12:41:31.078646 2018] [:error] [pid 15077] [client 142.157.144.81:54476] PHP Fatal error:  Uncaught DatabaseException: Insert statement did not execute successfully: Array\n(\n    [0] => HY000\n    [1] => 1366\n    [2] => Incorrect integer value: '' for column 'For_site' at row 1\n)\n in /var/www/LORIS/Loris/php/libraries/Database.class.inc:341\nStack trace:\n#0 /var/www/LORIS/Loris/php/libraries/Database.class.inc(243): Database->_realinsert('document_reposi...', Array, true)\n#1 /var/www/LORIS/Loris/modules/document_repository/ajax/documentEditUpload.php(94): Database->insert('document_reposi...', Array)\n#2 /var/www/LORIS/Loris/htdocs/AjaxHelper.php(104): require('/var/www/LORIS/...')\n#3 {main}\n  thrown in /var/www/LORIS/Loris/php/libraries/Database.class.inc on line 341, referer: http://132.206.37.24/document_repository/?reset=true
```

johnsaigle commented:
@ZainVirani I wasn't able to reproduce your error. 

What other parameters did you enter for the upload form? 

Does this error appear only when you're checked out on my branch?

johnsaigle commented:
> I'm getting lots of undefined index (openAccordian, filtered) and undefined variable (fileCategories) in my error log

I've addressed `openAccordion` and `filtered`. #3632 seems to be addressing some of these issues as well as `fileCategories`.

ZainVirani commented:
@johnsaigle sorry about late response
Category: text>test
Site: Any
File: test.txt
Tested again on your branch and failed
Tested on minor and it worked

Lemme review some code and see if i can spot anything

ZainVirani commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +44,39 @@
-        if (!file_exists($base_path . $puser)) {
-            mkdir($base_path . $puser, 0777);
+        $category   = $_POST['category']; // required
+        $site       = $_POST['site'] ?? ''; // required but allowed to be empty
```
@johnsaigle the error is occurring here
`SQLSTATE[HY000]: General error: 1366 Incorrect integer value: '' for column 'For_site' at row 1`

ZainVirani replied:
I think it should be `$_POST['site'] ?? null`

johnsaigle commented:
@ZainVirani I made the change you suggested. Please review again when you can

HenriRabalais commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -14,35 +14,27 @@
 ) {
-    header("HTTP/1.1 403 Forbidden");
+    http_response_code(403);
     exit;
```
Should this have an exit message to let the user know that they do not have permission to access the page, or is that implied?

johnsaigle replied:
That's what error 403 means, so yes. This function also automatically gives a response message (i.e. "Forbidden" in this case)

zaliqarosli commented:
you killed the tree structure haha

<img width="1280" alt="screenshot 2018-06-11 16 49 38" src="https://user-images.githubusercontent.com/17415878/41256406-7fdcdbc4-6d97-11e8-8c88-9c13224d392a.png">

johnsaigle commented:
@zaliqarosli :sweat_smile: thanks. Fixed!

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +44,39 @@
+            error_log(
+                'Creating document repository upload folder for ' .
+                'user ' . $name
+            );
```
from my understanding, is_writable returns true if the filename exists AND is writable. in the happy path case that !file_exists, $uploadPath can still be not writable. wouldn't this cause an error?

johnsaigle replied:
No I think the logic is good:
1. Check if file `is_writable`
2. If the file is NOT writable, then either: the file does not exist OR we have insufficient permissions
3. Check if the file exists
4. If the file does exist, but is not writable, give a forbidden error
5. If the file does not exist, create one.

zaliqarosli replied:
but what if the file does not exist AND we have insufficient permissions? wouldn't that case also enter the !is_writable($uploadPath) if statement?

johnsaigle replied:
>if the file does not exist AND we have insufficient permissions
`is_writable` checks one file/folder to see whether the current user (in this case, `www-data`/`apache`) has write permissions for that file. 
If the file doesn't exist it doesn't make sense to speak of what options are set on it.

If you're asking whether that user has permissions to write to `user_uploads/` that file is owned by the web user [by default](https://github.com/aces/Loris/blob/030d1c08989e7b94232d6996253bd8ffabb87199/tools/install.sh#L158).

This doesn't intersect in any way with the permissions associated with loris front-end users; it checks only whether the Linux system user can do file operations.

zaliqarosli replied:
i see, this makes sense

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -89,26 +86,21 @@
                  'File_size'     => $fileSize,
-                 'Data_dir'      => $fileBase,
-                 'uploaded_by'   => $puser,
+                 'Data_dir'      => $fullPath,
```
does changing this from relative to full path affect anywhere else where the `document_repository` table is used?

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -120,7 +112,10 @@
-        if (empty($category) && $category !== '0') {
+        // $category is a string representation of a number.
+        // only proceed if its int value is greater than or equal to 0
+        if (!is_numeric($category) || (int) $category < 0) {
```
its not clear to me what this is even with comments :(

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -120,7 +112,10 @@
+        // $category is a string representation of a number.
+        // only proceed if its int value is greater than or equal to 0
+        if (!is_numeric($category) || (int) $category < 0) {
+            http_response_code(403);
```
two different response codes?

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +44,39 @@
+        $uploadPath = $name . '/' . $fileName;
+        $fullPath   = __DIR__ . '/../user_uploads/' . $uploadPath;
+
+        if (!is_writable($f)) {
```
what is the $f ??

zaliqarosli commented:
when you want to edit a file, the error "category is required" even if one is selected from the drop down

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -89,26 +88,21 @@
                  'File_size'     => $fileSize,
-                 'Data_dir'      => $fileBase,
-                 'uploaded_by'   => $puser,
+                 'Data_dir'      => $name . $fileName,
```
i think you're missing a "/" in between the two variables

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -52,34 +44,41 @@
+        $fileName = $_FILES['file']['name'];
+        $fileType = pathinfo($_FILES['file']['name'], PATHINFO_EXTENSION);
+
+        $fullPath = __DIR__ . "/../user_uploads/$name/";
```
you can exchange for single quotes here if you want

johnsaigle replied:
In this case I think it's more clear with single quotes to show the directory structure. It's easy for me anyway than:

```
$fullPath = __DIR__ .  '/../user_uploads/' . $name . '/';
```

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -120,8 +114,10 @@
-            header("HTTP/1.1 400 Bad Request");
+        // $category is a string representation of a number.
+        // only proceed if its int value is greater than or equal to 0
+        if (intval($category) < 0) {
```
still not sure how checking that it is less than 0 is the same as checking its empty and not 0

ZainVirani replied:
I don't think it's equivalent for strings with an int value less than 0

johnsaigle replied:
@zaliqarosli I feel like this is better written this way. What we care about is the integer value being at least 0.

zaliqarosli replied:
@johnsaigle could you explain why we care about the integer value being atleast zero?

johnsaigle replied:
@zaliqarosli Just the way the module was coded. `File_category` in the DB is a value of 0 or higher that effectively represents a category ID. When uploading a file and choosing its category, what gets passed to the back-end isn't the name of the category but its ID. Do some poking around `document_repository` and `document_repository_categories` and you'll see what I mean.

Since $category should only be positive integers, it makes sense to validate this way.

zaliqarosli replied:
i think i understand what's going on now after looking into what $category actually is and how it's stored in the database. from what i see, category should be > 0 (`  `File_category` int(3) ~~unsigned~~ DEFAULT NULL ` ` for `document_repository` table)

johnsaigle replied:
0 is a valid unsigned variable though, no?

ZainVirani replied:
if category is always unsigned then my concern above is not valid 👍

zaliqarosli replied:
@johnsaigle yes it is valid, hence why i agree :)

driusan commented:
(Missing in the above summary: this also has a conflict.)

johnsaigle commented:
@driusan Thanks for the feedback. I've addressed the conflicts and made the changes you recommended. Please re-review when you can.

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -14,23 +14,15 @@
-require_once "Email.class.inc";
-$client = new NDB_Client();
-$client->initialize("../../project/config.xml");
 $factory = NDB_Factory::singleton();
```
$factory = \NDB_Factory::singleton()

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -14,23 +14,15 @@
+$name   = $userSingleton->getUsername();
 
 // create Database object
 $DB =& Database::singleton();
```
should use $DB = $factory->database();

zaliqarosli commented on modules/document_repository/ajax/documentEditUpload.php:
```diff
@@ -121,9 +125,13 @@
-            exit;
+        // $category is a string representation of an ID, and so should be at
+        // least equal to zero
+        if (intval($category) < 0) {
```
why does this need to be checked a second time? it appears above

johnsaigle replied:
Once for `edit` actions, another for `upload` actions. They are in different branches of the code.

zaliqarosli commented:
removed passed manual tests as it's been a while. happy to retest!

johnsaigle commented:
@zaliqarosli I made the changes you recommended. Please re-test :dancing_women:

zaliqarosli commented:
good to go! 👍 🌮

johnsaigle commented:
Only took 19 commits and 95 messages!
